org.wso2.carbon.identity.dao.stored.procedure/src/main/java/org/wso2/carbon/extension/identity/dao/stored/procedure/StoreProcedureBasedScopeClaimMappingDAOImpl.java

```diff
@@ -0,0 +1,111 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.extension.identity.dao.stored.procedure;
+
+import com.microsoft.sqlserver.jdbc.SQLServerCallableStatement;
+import com.microsoft.sqlserver.jdbc.SQLServerDataTable;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.core.util.IdentityDatabaseUtil;
+import org.wso2.carbon.identity.oauth.dto.ScopeDTO;
+import org.wso2.carbon.identity.oauth2.IdentityOAuth2Exception;
+import org.wso2.carbon.identity.oauth2.Oauth2ScopeConstants;
+import org.wso2.carbon.identity.openidconnect.dao.ScopeClaimMappingDAOImpl;
+
+import java.sql.CallableStatement;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.Types;
+import java.util.List;
+
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants.ARG_CLAIMS;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants.ARG_SCOPES;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants.ARG_TENANT_ID;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants
+        .CALL_INIT_SCOPE_CLAIM_MAPPING;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants.CLAIM;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants.CLAIM_ID;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants
+        .DBO_IDN_OAUTH_2_SCOPE;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants
+        .DBO_IDN_OIDC_SCOPE_CLAIMS;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants.DESCRIPTION;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants.DISPLAY_NAME;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants.NAME;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants.SCOPE_ID;
+import static org.wso2.carbon.extension.identity.dao.stored.procedure.StoreProcedureBasedDAOConstants.SCOPE_TYPE;
+
+/**
+ * Store Procedure Based OIDC Scope Claim mapping DAO.
+ */
+public class StoreProcedureBasedScopeClaimMappingDAOImpl extends ScopeClaimMappingDAOImpl {
+
+    private static final Log log = LogFactory.getLog(StoreProcedureBasedScopeClaimMappingDAOImpl.class);
+
+    @Override
+    public void initScopeClaimMapping(int tenantId, List<ScopeDTO> scopeClaimsList) throws IdentityOAuth2Exception {
+
+        Connection dbConnection = IdentityDatabaseUtil.getDBConnection(true);
```

Comment:
Do we have an interlock mechanism, if the scopes already created for the tenant?

Reply:
As we have the Database table level, we do need to have an interlock in the DAO level to check the existence of the scopes in the tenant creation? 

If the scope was already added (this is worst case as this will be tenant creation flow, only possible way is corrupted DB) then the stored procedure will fail with unique constrain violation. Do you see any other issues here?

Reply:
In the stored procedure level, we will have the interlock mechanism to handle the cases where the scopes are already created for the tenant. If any of the scope is already available for the tenant, the SP will roll back and throw the error to the server.